File: rrm_src/reasoning/WaveDynamics.ts
import { CognitiveEntity } from '../core/CognitiveEntity.js';
import { TensorVector, GLOBAL_DIMENSION } from '../core/config.js';
import { EntanglementOptimizer } from './EntanglementOptimizer.js';

/**
 * 🌊 WAVE DYNAMICS (Fase 4)
 * Menangani navigasi dan kausalitas (Entanglement) antar CognitiveEntities.
 * Menggunakan Hebbian Learning Matrix, Zero If-Else arsitektur, menghilangkan Map manual.
 */
export class WaveDynamics {
    private entanglementMatrix: Float32Array[] = [];
    private entitiesCache: CognitiveEntity[] = [];

    /**
     * Mendaftarkan seluruh agen/entitas dan menginisialisasi Matriks Entanglement
     */
    public initializeEntities(entities: CognitiveEntity[]): void {
        this.entitiesCache = entities;
        const numEntities = entities.length;

        // Buat N x N Float32Array untuk quantum coupling branchless tracking
        this.entanglementMatrix = new Array(numEntities);
        for (let i = 0; i < numEntities; i++) {
            this.entanglementMatrix[i] = new Float32Array(numEntities).fill(0.0);
            this.entanglementMatrix[i]![i] = 1.0; // Self-entangled = 1.0
        }
    }

    /**
     * Menjalankan Hebbian Learning: Otomatis mencari dan menyambungkan
     * "Neurons that fire together" tanpa if-else.
     */
    public evolveEntanglement(learningRate: number = 0.1): void {
        EntanglementOptimizer.optimize(this.entitiesCache, this.entanglementMatrix, learningRate);
    }

    /**
     * Memperbarui tensor inPlace berdasarkan daya pikat/tolak (Contrastive Update).
     * Mensimulasikan 'PhaseGradientOptimizer' lama.
     */
    private contrastiveUpdateInPlace(
        agentTensor: TensorVector,
        repulsorTensor: TensorVector,
        attractorTensor: TensorVector,
        alpha: number
    ): void {
        // Tarik ke arah attractor, tolak dari repulsor
        for (let i = 0; i < GLOBAL_DIMENSION; i++) {
            agentTensor[i] += alpha * (attractorTensor[i]! - repulsorTensor[i]!);
        }

        // L2 Normalization
        let magSq = 0;
        for (let i = 0; i < GLOBAL_DIMENSION; i++) {
            magSq += agentTensor[i]! * agentTensor[i]!;
        }
        const mag = Math.sqrt(magSq);

        // Zero-if branchless evaluation
        const isValidMag = mag > 1e-12;
        isValidMag && (() => {
            for (let i = 0; i < GLOBAL_DIMENSION; i++) agentTensor[i] /= mag;
        })();
    }

    /**
     * WAVE GRAVITY DRIVE (Huygens-Fresnel Navigation)
     * Mengkalkulasi pergerakan agen berdasarkan atraktor dan repulsor tensor di sekitarnya.
     */
    public applyWaveGravity(agent: CognitiveEntity, attractors: TensorVector[], repulsors: TensorVector[]): void {
        const totalAttractor = new Float32Array(GLOBAL_DIMENSION).fill(0);
        for (const attr of attractors) {
            for (let i = 0; i < GLOBAL_DIMENSION; i++) totalAttractor[i] += attr[i]!;
        }

        const totalRepulsor = new Float32Array(GLOBAL_DIMENSION).fill(0);
        for (const rep of repulsors) {
            for (let i = 0; i < GLOBAL_DIMENSION; i++) totalRepulsor[i] += rep[i]!;
        }

        // Terapkan medan ke tensor entitas
        this.contrastiveUpdateInPlace(agent.tensor, totalRepulsor, totalAttractor, 0.8);
    }

    /**
     * TRIGGER COLLAPSE
     * Meruntuhkan gelombang informasi ke agen-agen yang saling terikat (Entangled).
     * Jika A berubah, B terpengaruh SEBANDING dengan bobot entanglement-nya.
     */
    public triggerCollapse(sourceIndex: number): void {
        const sourceAgent = this.entitiesCache[sourceIndex];
        const numEntities = this.entitiesCache.length;

        // Zero if-else untuk safety check
        const isValidSource = !!sourceAgent && this.entanglementMatrix.length === numEntities;

        isValidSource && (() => {
            for (let targetIndex = 0; targetIndex < numEntities; targetIndex++) {
                // Skip self (branchless math logic)
                const isNotSelf = targetIndex !== sourceIndex;

                isNotSelf && (() => {
                    const targetAgent = this.entitiesCache[targetIndex]!;
                    const entanglementWeight = this.entanglementMatrix[sourceIndex]![targetIndex]!;

                    // Partial collapse berdasarkan coupling (0.0 to 1.0)
                    for(let i=0; i<GLOBAL_DIMENSION; i++){
                        // Interpolasi linear murni: W * Source + (1-W) * Target
                        targetAgent.tensor[i] = (entanglementWeight * sourceAgent!.tensor[i]!) + ((1.0 - entanglementWeight) * targetAgent.tensor[i]!);
                    }

                    // Mark entanglement status as combined weight
                    targetAgent.entanglement_status = Math.max(targetAgent.entanglement_status, entanglementWeight);
                })();
            }
        })();
    }
}

File: rrm_rust/src/reasoning/wave_dynamics.rs
use crate::core::config::{GLOBAL_DIMENSION, MAX_ENTITIES};
use crate::core::entity_manifold::EntityManifold;
use crate::reasoning::entanglement_optimizer::EntanglementOptimizer;
use ndarray::Array1;

/// 🌊 WAVE DYNAMICS (Fase 4)
/// OOP-Free Hebbian Learning Matrix & Huygens-Fresnel Navigation.
pub struct WaveDynamics {
    // 🏛️ HUKUM 8: Gunakan satu blok memori (N * N) agar L1 Cache tidak meleset
    entanglement_matrix: Vec<f32>,
}

impl WaveDynamics {
    pub fn new() -> Self {
        let mut matrix = vec![0.0; MAX_ENTITIES * MAX_ENTITIES];
        // Self-entanglement
        for i in 0..MAX_ENTITIES {
            matrix[i * MAX_ENTITIES + i] = 1.0;
        }
        Self {
            entanglement_matrix: matrix,
        }
    }

    /// Mendaftarkan active boundaries. Unlike TS which holds a ref, we pass manifold explicitly.
    pub fn initialize_entities(&mut self, manifold: &EntityManifold) {
        let count = manifold.active_count;

        // Reset matrix only for active boundaries
        for i in 0..count {
            let row_offset = i * MAX_ENTITIES;
            for j in 0..count {
                self.entanglement_matrix[row_offset + j] = 0.0;
            }
            self.entanglement_matrix[row_offset + i] = 1.0;
        }
    }

    /// Menjalankan Hebbian Learning di atas Structure of Arrays.
    pub fn evolve_entanglement(&mut self, manifold: &EntityManifold, learning_rate: f32) {
        EntanglementOptimizer::optimize(manifold, &mut self.entanglement_matrix, learning_rate);
    }

    /// Memperbarui tensor inPlace berdasarkan daya pikat/tolak (Contrastive Update).
    fn contrastive_update_in_place(
        agent_tensor: &mut Array1<f32>,
        repulsor_tensor: &Array1<f32>,
        attractor_tensor: &Array1<f32>,
        alpha: f32,
    ) {
        // Tarik ke arah attractor, tolak dari repulsor
        for i in 0..GLOBAL_DIMENSION {
            agent_tensor[i] += alpha * (attractor_tensor[i] - repulsor_tensor[i]);
        }

        // L2 Normalization
        let mut mag_sq = 0.0;
        for i in 0..GLOBAL_DIMENSION {
            mag_sq += agent_tensor[i] * agent_tensor[i];
        }
        let mag = mag_sq.sqrt();

        // Menggunakan EPSILON agar tidak perlu if (mag > 0)
        let inv_mag = 1.0 / (mag + 1e-15);
        for i in 0..GLOBAL_DIMENSION {
            agent_tensor[i] *= inv_mag;
        }
    }

    /// WAVE GRAVITY DRIVE (Huygens-Fresnel Navigation)
    /// Mengkalkulasi pergerakan agen berdasarkan atraktor dan repulsor tensor di sekitarnya.
    pub fn apply_wave_gravity(
        &self,
        manifold: &mut EntityManifold,
        agent_index: usize,
        attractors: &[&Array1<f32>],
        repulsors: &[&Array1<f32>],
    ) {
        let mut total_attractor = Array1::<f32>::zeros(GLOBAL_DIMENSION);
        for attr in attractors {
            for i in 0..GLOBAL_DIMENSION {
                total_attractor[i] += attr[i];
            }
        }

        let mut total_repulsor = Array1::<f32>::zeros(GLOBAL_DIMENSION);
        for rep in repulsors {
            for i in 0..GLOBAL_DIMENSION {
                total_repulsor[i] += rep[i];
            }
        }

        let mut agent_tensor = manifold.get_spatial_tensor(agent_index);
        Self::contrastive_update_in_place(&mut agent_tensor, &total_repulsor, &total_attractor, 0.8);

        let mut t_mut = manifold.get_spatial_tensor_mut(agent_index);
        for i in 0..GLOBAL_DIMENSION {
            t_mut[i] = agent_tensor[i];
        }
    }

    /// TRIGGER COLLAPSE
    /// Meruntuhkan gelombang informasi ke agen-agen yang saling terikat (Entangled).
    pub fn trigger_collapse(&self, manifold: &mut EntityManifold, source_index: usize) {
        let num_entities = manifold.active_count;

        let source_tensor = manifold.get_spatial_tensor(source_index).to_owned();
        let source_row_offset = source_index * MAX_ENTITIES;

        for target_index in 0..num_entities {
            if target_index == source_index || manifold.masses[target_index] == 0.0 {
                continue;
            }

            let entanglement_weight = self.entanglement_matrix[source_row_offset + target_index];

            {
                let mut target_tensor = manifold.get_spatial_tensor_mut(target_index);
                for i in 0..GLOBAL_DIMENSION {
                    target_tensor[i] = (entanglement_weight * source_tensor[i])
                        + ((1.0 - entanglement_weight) * target_tensor[i]);
                }
            }

            let current_status = manifold.entanglement_status[target_index];
            if entanglement_weight > current_status {
                manifold.entanglement_status[target_index] = entanglement_weight;
            }
        }
    }
}
